lib/contacts_service.dart

```diff
@@ -46,6 +46,21 @@ class ContactsService {
     return contacts.map((m) => Contact.fromMap(m));
   }
 
+  /// Fetches the avatar for the contacts with the provided [identifiers].
+  /// Returns a list of avatars in the same order as the [identifiers] given.
+  /// If a user does not have an avatar, then `null` is returned in that slot.
+  static Future<Iterable<Uint8List>> getAvatarForContacts(
```

Comment:
API Suggestion: This kind of index coupling is error prone.

I would propose that you implement:

```dart
Future<Iterable<Contact>> fillAvatars(final Iterable<Contact> contacts);
```

The client code will be cleaner and less error prone.

Reply:
It seems like Contact is not immutable :-/. It supports (+) operator to fill in missing fields so you can literally implement:

```dart
Future<void> fillAvatars(final Iterable<Contact> contacts);
```

I disagree with this but we also stay consistent to the repo. Implementing immutable Contact objects would require us to rewrite the whole API.

Reply:
Yeah, I was grappling with the same problem. I wanted to do an immutable transform like you suggested, but the mutability of the `Contact` class makes that a little weird.

Regarding:

```dart
Future<void> fillAvatars(final Iterable<Contact> contacts);
```

I had a similar idea initially, and debated it with myself for quite a while. My one concern with that approach is that it **requires** clients to use the `Contact` class in a mutable fashion, which they may not want to do. If they wanted to treat it immutability and had this API, the best they could do is use `Contact#toMap()` and `Contact.fromMap()` to clone it and then call `fillAvatars()` on the clones.

I'm thinking we could provide both APIs, one called `getAvatars()` which just returns the avatars like we already have here, and the other called `fillAvatars()` which populates `Contact#avatar` (probably by wrapping `getAvatars()` under the hood). I think this is a little more flexible to what clients might want, though the downside is that it is more code/API surface to maintain, and it also exposes a slightly more complicated API surface to clients (`getAvatars()` and `fillAvatars()` do the same thing, just in different ways).

Any thoughts/concerns about supporting both forms of use?

Reply:
Great points. I would not want to crowd the API so supporting both is not a good option.

Let's consider the use cases. If the client wanted *all* the avatars, they would have asked for them when they called `getContacts()`. So they either want a small subset or a single avatar.

Let's go with option #3: Provide only one API for getting a single Avatar.

```dart
Future<Uint8List> getAvatar(Contact contact);
```

If a client wants to populate multiple avatars all they need is:

```dart
List<Uint8List> avatars = await Future.wait(
  getAvatar(myContact1),
  getAvatar(myContact2),
  getAvatar(myContact3),
);
```

This has the advantage of spawning N background tasks to grab N avatars. I just made a comment on capping the number of threads by using a custom executor such as ThreadPoolExecutor.

This also shifts the responsibility of maintaining the indexing to the client as Future.wait() will simply return a list of results in the same order of the method calls.

Reply:
> Let's consider the use cases. If the client wanted _all_ the avatars, they would have asked for them when they called `getContacts()`. So they either want a small subset or a single avatar.

I don't have full context as to the immediate need for this feature, though my interpretation of the most common case would be a client who wants to display a contact list as quickly as possible, and then lazy load **all** the contacts for that list. That use case would not want to include thumbnails in their call to `getContacts()`. You are right that some clients might want only a subset, such as the 10 or so contacts which happen to be currently visible on screen and lazy-loading as the user scrolls, so we should support that as well.

> Let's go with option #3: Provide only one API for getting a single Avatar.
> 
> ```dart
> Future<Uint8List> getAvatar(Contact contact);
> ```
> 
> If a client wants to populate multiple avatars all they need is:
> 
> ```dart
> List<Uint8List> avatars = await Future.wait(
>   getAvatar(myContact1),
>   getAvatar(myContact2),
>   getAvatar(myContact3),
> );
> ```

This is definitely a simpler API, with a great advantage that the contacts can be updated individually, rather than waiting for all of them before painting any one. It also means that these avatars will be loaded in parallel.

```dart
await Future.wait(
  getAvatar(myContact1).then((avatar) => setState(() => myContact1.avatar = avatar)),
  getAvatar(myContact2).then((avatar) => setState(() => myContact2.avatar = avatar)),
  getAvatar(myContact3).then((avatar) => setState(() => myContact3.avatar = avatar)),
);
```

> This has the advantage of spawning N background tasks to grab N avatars. I just made a comment on capping the number of threads by using a custom executor such as ThreadPoolExecutor.

I had previously ruled out this kind of API specifically *because* it would take N background tasks and N method channel invocations. I'm not too familiar with the various executors and their caveats as well as the performance implications of method channel, so is that a reasonable overhead to impose? If this is reasonably performant, then I definitely agree this is the simplest and most flexible API design.

Reply:
That's a reasonable overhead. It is also easily configurable by the app. If they are trying to load 100 avatars at once, then they will pay some performance penalty but that's exactly why we are exposing this method in a flexible way so they *don't* do that. Avatars, while unneeded for the UI to function, are still important. Without them, the app would look broken.

We should restrict threads to 10. We might want to make that user configurable but it could be a foot gun in the hands of users with not much Android experience (majority of Flutter users). So let's make that decision for them.

Reply:
Thanks for confirming the performance implications. Updated the API to `getAvatar()` like you suggested.

I definitely agree that letting users configure the executor is a great way for them to shoot themselves in the foot, especially Flutter developers who don't deal with Android-specific concerns on a day-to-day basis. I left out any configuration for now, and I figure it can be added in later if someone *really* wants that feature.